src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnection.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.SignalR;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.Azure.SignalR.Protocol;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.SignalR.AspNet
+{
+    internal partial class ServiceConnection : ServiceConnectionBase
+    {
+        private readonly ConcurrentDictionary<string, AzureTransport> _clientConnections = new ConcurrentDictionary<string, AzureTransport>(StringComparer.Ordinal);
+
+        private readonly string _hubName;
+        private readonly IConnectionFactory _connectionFactory;
+        private readonly IClientConnectionManager _clientConnectionManager;
+
+        public ServiceConnection(
+            string hubName,
+            string connectionId,
+            IServiceProtocol serviceProtocol,
+            IConnectionFactory connectionFactory,
+            IClientConnectionManager clientConnectionManager,
+            ILogger logger)
+            : base(serviceProtocol, logger, connectionId)
+        {
+            _hubName = hubName;
+            _connectionFactory = connectionFactory;
+            _clientConnectionManager = clientConnectionManager;
+        }
+
+        protected override Task<ConnectionContext> CreateConnection()
+        {
+            return _connectionFactory.ConnectAsync(TransferFormat.Binary, _connectionId, _hubName);
+        }
+
+        protected override Task DisposeConnection()
+        {
+            var connection = _connection;
+            _connection = null;
+            return _connectionFactory.DisposeAsync(connection);
+        }
+
+        protected override Task CleanupConnections()
+        {
+            try
+            {
+                foreach (var connectionId in _clientConnections.Keys)
```

Comment:
Performance tip: It's more efficient to loop over the ConcurrentDictionary itself rather than keys or values. Each of those collections takes a lock on all of the buckets.